dialog/color.go

```diff
@@ -0,0 +1,334 @@
+		// Add divider and recents if there are any
+		contents = append(contents, canvas.NewLine(theme.ShadowColor()), recent)
+	}
+	return
+}
+
+func (p *ColorPickerDialog) selectColor(c color.Color) {
+	p.dialog.Hide()
+	writeRecentColor(colorToString(p.color))
+	if f := p.callback; f != nil {
+		f(c)
+	}
+}
+
+func clamp(value, min, max int) int {
+	if value < min {
+		value = min
+	}
+	if value > max {
+		value = max
+	}
+	return value
+}
+
+func wrapHue(hue int) int {
+	for hue < 0 {
+		hue += 360
+	}
+	for hue > 360 {
+		hue -= 360
+	}
+	return hue
+}
+
+func newColorButtonBox(colors []color.Color, icon fyne.Resource, callback func(color.Color)) fyne.CanvasObject {
+	var objects []fyne.CanvasObject
+	if icon != nil && len(colors) > 0 {
+		objects = append(objects, widget.NewIcon(icon))
+	}
+	for _, c := range colors {
+		objects = append(objects, newColorButton(c, callback))
+	}
+	return fyne.NewContainerWithLayout(layout.NewGridLayoutWithColumns(9), objects...)
+}
+
+func newCheckeredBackground() *canvas.Raster {
+	return canvas.NewRasterWithPixels(func(x, y, _, _ int) color.Color {
+		const boxSize = 10
+
+		if (x/boxSize)%2 == (y/boxSize)%2 {
+			return color.Gray{Y: 58}
+		}
+
+		return color.Gray{Y: 84}
+	})
+}
+
+const (
+	preferenceRecents    = "color_recents"
+	preferenceMaxRecents = 8
+)
+
+func readRecentColors() (recents []string) {
+	for _, r := range strings.Split(fyne.CurrentApp().Preferences().String(preferenceRecents), ",") {
+		if r != "" {
+			recents = append(recents, r)
+		}
+	}
+	return
+}
+
+func writeRecentColor(color string) {
+	recents := []string{color}
+	for _, r := range readRecentColors() {
+		if r == color {
+			continue // Color already in recents
+		}
+		recents = append(recents, r)
+	}
+	if len(recents) > preferenceMaxRecents {
+		recents = recents[:preferenceMaxRecents]
+	}
+	fyne.CurrentApp().Preferences().SetString(preferenceRecents, strings.Join(recents, ","))
+}
+
+func colorToString(c color.Color) string {
+	red, green, blue, alpha := colorToRGBA(c)
+	if alpha == 0xff {
+		return fmt.Sprintf("#%02x%02x%02x", red, green, blue)
+	}
+	return fmt.Sprintf("#%02x%02x%02x%02x", red, green, blue, alpha)
+}
+
+func stringToColor(s string) (color.Color, error) {
+	var c color.NRGBA
+	var err error
+	if len(s) == 7 {
+		c.A = 0xFF
+		_, err = fmt.Sscanf(s, "#%02x%02x%02x", &c.R, &c.G, &c.B)
+	} else {
+		_, err = fmt.Sscanf(s, "#%02x%02x%02x%02x", &c.R, &c.G, &c.B, &c.A)
+	}
+	return c, err
+}
+
+func stringsToColors(ss ...string) (colors []color.Color) {
+	for _, s := range ss {
+		if s == "" {
+			continue
+		}
+		c, err := stringToColor(s)
+		if err != nil {
+			fyne.LogError("Couldn't parse color:", err)
+		} else {
+			colors = append(colors, c)
+		}
+	}
+	return
+}
+
+func colorToHSLA(c color.Color) (int, int, int, int) {
+	r, g, b, a := colorToRGBA(c)
+	h, s, l := rgbToHsl(r, g, b)
+	return h, s, l, a
+}
+
+func colorToRGBA(c color.Color) (r, g, b, a int) {
```

Comment:
This function seems like it could be simplified.
You could resolve your TODO and improve this at the same time.
Just call .RGBA() on any colour.Color and then divide by the alpha value (if non-1 and non-0)

Reply:
Okay cool, that did simplify it, thanks. I did need to add two cases for NRGBA tho which seems odd - I got an error that it was expecting *color.NRGBA but was actually color.NRGBA :/